drivers/led/is31fl3733.c

```diff
@@ -0,0 +1,449 @@
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	/* Write LED on/off states */
+	ret = is31fl3733_select_page(dev, CMD_SEL_LED);
+	if (ret < 0) {
+		return ret;
+	}
+	LOG_HEXDUMP_DBG(on_start, (on_cursor - on_start), "LED on states");
+	ret = i2c_write_dt(&config->bus, on_start, (on_cursor - on_start));
+	if (ret < 0) {
+		return ret;
+	}
+	/* Write LED PWM states */
+	ret = is31fl3733_select_page(dev, CMD_SEL_PWM);
+	if (ret < 0) {
+		return ret;
+	}
+	LOG_HEXDUMP_DBG(pwm_start, (pwm_cursor - pwm_start), "PWM states");
+	ret = i2c_write_dt(&config->bus, pwm_start, (pwm_cursor - pwm_start));
+	return ret;
+}
+
+static int is31fl3733_init(const struct device *dev)
+{
+	const struct is31fl3733_config *config = dev->config;
+	struct is31fl3733_data *data = dev->data;
+	int ret = 0U;
+	uint8_t dummy;
+
+	if (!i2c_is_ready_dt(&config->bus)) {
+		LOG_ERR("I2C device not ready");
+		return -ENODEV;
+	}
+	if ((config->sdb.port != NULL) && !gpio_is_ready_dt(&config->sdb)) {
+		LOG_ERR("GPIO SDB pin not ready");
+		return -ENODEV;
+	}
+
+	ret = is31fl3733_select_page(dev, CMD_SEL_FUNC);
+	if (ret < 0) {
+		return ret;
+	}
+	/*
+	 * read reset reg to reset all registers to POR state,
+	 * in case we are booting from a warm reset.
+	 */
+	ret = i2c_reg_read_byte_dt(&config->bus, RESET_REG, &dummy);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Clear LED state and PWM state buffers, to reflect led driver state */
+	memset(data->led_state, 0, sizeof(data->led_state));
+	memset(data->pwm_state, 0, sizeof(data->pwm_state));
+
+
+	if (config->sdb.port != NULL) {
+		/* Set SDB pin high to exit hardware shutdown */
+		ret = gpio_pin_configure_dt(&config->sdb, GPIO_OUTPUT);
+		if (ret < 0) {
+			return ret;
+		}
+		ret = gpio_pin_set_dt(&config->sdb, 1);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+	ret = is31fl3733_select_page(dev, CMD_SEL_FUNC);
+	if (ret < 0) {
+		return ret;
+	}
+	/* Set global current control register based off devicetree value */
+	ret = i2c_reg_write_byte_dt(&config->bus, GLOBAL_CURRENT_CTRL_REG,
+				config->current_limit);
+	if (ret < 0) {
+		return ret;
+	}
+	/* As a final step, we exit software shutdown, disabling display
+	 * blanking. We also set the LED controller sync mode here.
+	 */
+	ret = is31fl3733_select_page(dev, CMD_SEL_FUNC);
+	if (ret < 0) {
+		return ret;
+	}
+	data->conf_reg = (config->sync << CONF_REG_SYNC_SHIFT) | CONF_REG_SSD_MASK;
+	ret = i2c_reg_write_byte_dt(&config->bus, CONF_REG, data->conf_reg);
+	if (ret < 0) {
+		return ret;
+	}
+	return ret;
+}
+
+/* Custom IS31FL3733 specific APIs */
+
+/**
+ * @brief Blanks IS31FL3733 LED display.
+ *
+ * When blank_en is set, the LED display will be disabled. This can be used for
+ * flicker-free display updates, or power saving.
+ *
+ * @param dev: LED device structure
+ * @param blank_en: should blanking be enabled
+ * @return 0 on success, or negative value on error.
+ */
+int is31fl3733_blank(const struct device *dev, bool blank_en)
```

Comment:
My concern with this approach is that if the application enables power management with CONFIG_PM, they will be required to enable device level power management as well unless they want the LED controller to automatically blank every time the system enters into a low power state. I'm not sure it's a safe assumption that every system wants the LED controller to automatically blank when entering low power modes.

Reply:
It would be nice if it was possible for an application to disable PM for a device/driver.

In https://github.com/zephyrproject-rtos/zephyr/pull/58466 @Manimaran-A is proposing to introduce a DT flag to enable/disable PM support for a LED driver. I don't think this is correct but I don't have a better idea...

Please I'd like your opinion on this.